relay-server/server/relayServer.go

```diff
@@ -33,9 +33,14 @@ var Running bool
 // ClientList Map
 var ClientList map[string]int
 
+// ClientList Lock
+var ClientListLock *sync.RWMutex
```

Comment:
```suggestion
var ClientListLock *sync.Mutex
```

Reply:
> while testing we found we still had the above mentioned issue
> 
> the problem still persisted even after adding lock because we were using RWMutex because it is a type of lock that allows multiple readers to access the shared resource simultaneously
> 
> Also in GetFeedsFromNodes() function do this add lock above `if _, ok := ClientList[nodeIP]; !ok ` and unlock after that like this
> 
> ```
> 			for _, nodeIP := range newNodes {
> 				ClientListLock.Lock()
> 				if _, ok := ClientList[nodeIP]; !ok {
> 					ClientList[nodeIP] = 1
> 					go connectToKubeArmor(nodeIP, rs.Port)
> 				}
> 				ClientListLock.Unlock()
> 			}
> 
> 			time.Sleep(time.Second * 1)
> 		}
> ```

IMO we could use ```ClientListLock.RLock()```  instead of ```ClientListLock.Lock()``` Since just a reading function won't be changing any content, so it is acceptable to allow other readers to read at the same time to enhance the program’s performance.